src/rgw/rgw_object_expirer_core.cc

```diff
@@ -52,11 +52,14 @@ int RGWObjectExpirer::init_bucket_info(const string& tenant_name,
    * punching the tenant through the objexp_hint_entry, but now we
    * find that our instances do not actually have tenants. They are
    * unique thanks to IDs. So the tenant string is not needed...
+
+   * XXX reloaded: it turns out tenants were needed after all since bucket ids
+   * are ephemeral, good call encoding tenant info!
    */
-  const string bucket_instance_id = bucket_name + ":" + bucket_id;
-  int ret = store->get_bucket_instance_info(obj_ctx, bucket_instance_id,
-          bucket_info, NULL, NULL);
-  return ret;
+
+  return store->get_bucket_info(obj_ctx, tenant_name, bucket_name,
+				bucket_info, nullptr, nullptr);
+
```

Comment:
this does seem to resolve the reshard case, but wouldn't catch the case where the bucket was deleted and recreated

to delete the bucket, we would have first deleted any objects in the original bucket with a delete_at attribute - does that guarantee the removal of associated objexp_hint_entries?

Reply:
Not really, but we do pass on the delete_at attr down to rados which again checks the xattr before proceeding afaik. Good catch, I'll see what happens in this case

Reply:
Current code does seem to behave correctly for the following 2 cases:
1. Bucket recreated, object recreated with no more delete_at attr
2. Bucket recreated, object recreated with a longer delete_at attr